src/xdp/xsk.c

```diff
@@ -2947,52 +3082,50 @@ XskSockoptSetUmem(
-        Status = STATUS_INSUFFICIENT_RESOURCES;
-        goto Exit;
-    }
+    if (RequestorMode != KernelMode) {
+        Umem->Mapping.Mdl =
+            IoAllocateMdl(
+                Umem->Reg.Address,
+                (ULONG)Umem->Reg.TotalSize,
+                FALSE, // SecondaryBuffer
+                FALSE, // ChargeQuota
+                NULL); // Irp
+        if (Umem->Mapping.Mdl == NULL) {
+            Status = STATUS_INSUFFICIENT_RESOURCES;
+            goto Exit;
+        }
 
-    __try {
-        MmProbeAndLockPages(Umem->Mapping.Mdl, RequestorMode, IoWriteAccess);
-    } __except (EXCEPTION_EXECUTE_HANDLER) {
-        Status = GetExceptionCode();
-        goto Exit;
-    }
+        __try {
+            MmProbeAndLockPages(Umem->Mapping.Mdl, RequestorMode, IoWriteAccess);
+        } __except (EXCEPTION_EXECUTE_HANDLER) {
+            Status = GetExceptionCode();
+            goto Exit;
+        }
 
-    //
-    // MmGetSystemAddressForMdlSafe and MmMapLockedPagesSpecifyCache do not
-    // preserve large pages in system address mappings. Use the reserved MDL
-    // mapping routines to ensure large pages are propagated into the kernel.
-    // Note that the reserved mapping allocates the mapping size based on best-
-    // case page-aligned buffers, so account for the MDL offset, too.
-    //
-    Umem->ReservedMapping =
-        MmAllocateMappingAddress(
-            BYTE_OFFSET(Umem->Reg.Address) + Umem->Reg.TotalSize, POOLTAG_UMEM);
-    if (Umem->ReservedMapping == NULL) {
-        Status = STATUS_INSUFFICIENT_RESOURCES;
-        goto Exit;
-    }
+        //
+        // MmGetSystemAddressForMdlSafe and MmMapLockedPagesSpecifyCache do not
+        // preserve large pages in system address mappings. Use the reserved MDL
+        // mapping routines to ensure large pages are propagated into the kernel.
+        // Note that the reserved mapping allocates the mapping size based on best-
+        // case page-aligned buffers, so account for the MDL offset, too.
+        //
+        Umem->ReservedMapping =
+            MmAllocateMappingAddress(
+                BYTE_OFFSET(Umem->Reg.Address) + Umem->Reg.TotalSize, POOLTAG_UMEM);
+        if (Umem->ReservedMapping == NULL) {
+            Status = STATUS_INSUFFICIENT_RESOURCES;
+            goto Exit;
+        }
 
-    Umem->Mapping.SystemAddress =
-        MmMapLockedPagesWithReservedMapping(
-            Umem->ReservedMapping, POOLTAG_UMEM, Umem->Mapping.Mdl, MmCached);
-    if (Umem->Mapping.SystemAddress == NULL) {
-        Status = STATUS_INSUFFICIENT_RESOURCES;
-        goto Exit;
+        Umem->Mapping.SystemAddress =
+            MmMapLockedPagesWithReservedMapping(
+                Umem->ReservedMapping, POOLTAG_UMEM, Umem->Mapping.Mdl, MmCached);
+        if (Umem->Mapping.SystemAddress == NULL) {
+            Status = STATUS_INSUFFICIENT_RESOURCES;
+            goto Exit;
+        }
+    } else {
+        Umem->Mapping.SystemAddress = Umem->Reg.Address;
```

Comment:
Should we do an assert here that the Umem address is a system address, not a user mode address?

Reply:
Added assert